src/Diagnostics/CodeAnalysis/Core/MetaAnalyzers/RegisterActionAnalyzer.cs

```diff
@@ -110,32 +154,104 @@ protected abstract class RegisterActionCompilationAnalyzer : SyntaxNodeWithinAna
+            private HashSet<IParameterSymbol> _declaredStartAnalysisContextParams;
+
+            /// <summary>
+            /// Set of declared start analysis context parameters that need to be skipped for <see cref="StartActionWithNoRegisteredActionsRule"/> and <see cref="StartActionWithOnlyEndActionRule"/>.
+            /// This is to avoid false positives where context types are passed as arguments to a different invocation, and hence the registration responsibility is not on the current method.
+            /// </summary>
+            private HashSet<IParameterSymbol> _startAnalysisContextParamsToSkip;
+
+            protected RegisterActionCodeBlockAnalyzer(
                 INamedTypeSymbol analysisContext,
                 INamedTypeSymbol compilationStartAnalysisContext,
                 INamedTypeSymbol codeBlockStartAnalysisContext,
                 INamedTypeSymbol symbolKind,
                 INamedTypeSymbol diagnosticAnalyzer,
                 INamedTypeSymbol diagnosticAnalyzerAttribute)
-                : base(diagnosticAnalyzer, diagnosticAnalyzerAttribute)
             {
                 _analysisContext = analysisContext;
                 _compilationStartAnalysisContext = compilationStartAnalysisContext;
                 _codeBlockStartAnalysisContext = codeBlockStartAnalysisContext;
                 _symbolKind = symbolKind;
+
+                _nestedActionsMap = null;
+                _declaredStartAnalysisContextParams = null;
+                _startAnalysisContextParamsToSkip = null;
             }
 
             protected abstract IEnumerable<SyntaxNode> GetArgumentExpressions(TInvocationExpressionSyntax invocation);
+            protected abstract SyntaxNode GetArgumentExpression(TArgumentSyntax argument);
             protected abstract SyntaxNode GetInvocationExpression(TInvocationExpressionSyntax invocation);
+            protected abstract SyntaxNode GetInvocationReceiver(TInvocationExpressionSyntax invocation);
             protected abstract bool IsSyntaxKind(ITypeSymbol type);
+            protected abstract TLanguageKindEnum InvocationExpressionKind { get; }
+            protected abstract TLanguageKindEnum ArgumentSyntaxKind { get; }
+            protected abstract TLanguageKindEnum ParameterSyntaxKind { get; }
+
+            internal void CodeBlockStartAction(CodeBlockStartAnalysisContext<TLanguageKindEnum> codeBlockContext)
+            {
+                var method = codeBlockContext.OwningSymbol as IMethodSymbol;
+                if (!ShouldAnalyze(method))
+                {
+                    return;
+                }
+
+                foreach (var param in method.Parameters)
+                {
+                    AnalyzeParameterDeclaration(param);
+                }
+
+                codeBlockContext.RegisterSyntaxNodeAction(AnalyzeInvocation, InvocationExpressionKind);
+                codeBlockContext.RegisterSyntaxNodeAction(AnalyzerParameterSyntax, ParameterSyntaxKind);
```

Comment:
It's primarily for lambda parameters, the above case just handles method declaration parameters.